lib/ordered_set.h

```diff
@@ -17,234 +17,540 @@ limitations under the License.
+
+          T  get_payload () const          { return _t; }
+};
+
+/*
+template <typename T>
+bool operator== (const EnlistedNode <T> &x, const EnlistedNode <T> &y)
+noexcept (noexcept (x.ref_payload () == y.ref_payload ()))
+{
+    return x.ref_payload () == y.ref_payload ();
+}
+
+template <typename T>
+bool operator!= (const EnlistedNode <T> &x, const EnlistedNode <T> &y)
+noexcept (noexcept (x.ref_payload () != y.ref_payload ()))
+{
+    return x.ref_payload () != y.ref_payload ();
+}
+
+template <typename T>
+bool operator<  (const EnlistedNode <T> &x, const EnlistedNode <T> &y)
+noexcept (noexcept (x.ref_payload () <  y.ref_payload ()))
+{
+    return x.ref_payload () <  y.ref_payload ();
+}
+
+template <typename T>
+bool operator>  (const EnlistedNode <T> &x, const EnlistedNode <T> &y)
+noexcept (noexcept (x.ref_payload () >  y.ref_payload ()))
+{
+    return x.ref_payload () >  y.ref_payload ();
+}
+
+template <typename T>
+bool operator<= (const EnlistedNode <T> &x, const EnlistedNode <T> &y)
+noexcept (noexcept (x.ref_payload () <= y.ref_payload ()))
+{
+    return x.ref_payload () <= y.ref_payload ();
+}
+
+template <typename T>
+bool operator>= (const EnlistedNode <T> &x, const EnlistedNode <T> &y)
+noexcept (noexcept (x.ref_payload () >= y.ref_payload ()))
+{
+    return x.ref_payload () >= y.ref_payload ();
+}
+*/
+
+
+//============================================================================
+// EnlistedComp <Comp>:
+//
+// If Comp is a Strict Weak Ordering
+// for comparing objects of type T,
+// then EnlistedComp <Comp> is a Strict Weak Ordering
+// for comparing object of type EnlistedNode <T>.
+//============================================================================
+
+template <typename Comp>
+class EnlistedComp
+{
+public:
+    typedef EnlistedNode <typename Comp::first_argument_type>  first_argument_type;
+    typedef EnlistedNode <typename Comp::second_argument_type> second_argument_type;
+    typedef               typename Comp::result_type           result_type;
+
+private:
+    Comp _comp;
+
+public:
+    EnlistedComp () = default;
+
+    explicit EnlistedComp (const Comp &comp):
+        _comp (comp)
+    {}
+
+    result_type operator() (const first_argument_type &x, const second_argument_type &y) const
+    noexcept (noexcept (_comp (x.ref_payload (), y.ref_payload ())))
+    {
+        return _comp (x.ref_payload (), y.ref_payload ());
+    }
+};
+
+
+//============================================================================
+// ordered_set:
+//============================================================================
+
+template
+<
+    typename T,
+    class    Comp  = std::less <T>,
+    class    Alloc = std::allocator <T>
+>
+class ordered_set
+{
+// The (non-intrusive) tree:
+
+private:
+    typedef
+        EnlistedNode <T>
+        MyEnlistedNode;
+
+    typedef
+        EnlistedComp <Comp>
+        MyEnlistedComp;
+
+    typedef
+        typename Alloc::template rebind <MyEnlistedNode>::other
```

Comment:
rebind is deprecated and removed in some newer compilers -- need to use `rebind_alloc` instead.